### PR TITLE
Multi user file browsing, closes #2479

### LINKED ISSUE #2479: Multiuser support: Filesystem operations malfunction when using the app in secondary users
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**. Not a [question or feature request](https://github.com/gsantner/markor/issues/new/choose).
- [X] The topic is **not** already reported at [Issues](https://github.com/gsantner/markor/issues?q=#js-issues-search). _(I've searched it)_.
- [X] Markor **is** up to date. See [Releases](https://github.com/gsantner/markor/tags) for the latest version. Updates are available from [F-Droid](https://f-droid.org/en/packages/net.gsantner.markor/) and GitHub.
- [X] The bug is still present in the latest development version (git master). (Please [download](https://nightly.link/gsantner/markor/workflows/build-android-project/master) and try the test version of Markor, named **Marder**. Don't worry; Markor and Marder appear as completely separate applications. You can install both side-by-side, and Markor settings are not touched. In case the issue is resolved there, you don't need to create a bug report. The change will be part of the next Markor update.)

### Description

Not to be patronizing, but I thought this issue could be easy to misunderstand, so I just wanted to state what I mean by the terminology I use (derived from [here](https://source.android.com/docs/devices/admin/multi-user)):
* System user, meaning the user that must exist on single-user phones; the primary user. On my phone, its data is stored in /storage/emulated/0/.
* Secondary user, meaning an additional user besides the owner. On my phone, the secondary user I use Markor in has its data stored in /storage/emulated/10/.
* Symlinks, meaning those shortcuts to certain parts of the filesystem that can be seen if you tap on the "Go to" folder button at the right of the top navigation bar. Alternatively, you can find the symlinks by going up directories until you reach "storage".

***

When using Markor in a secondary user, navigation through folders malfunctions. Also, copy/move operations to symlinks malfunction. This may be related to #2350. Some specifics:

1. In the filesystem browser, when tapping "..", Markor will usually go all the way up to "storage" and show the symlinks. I expected Markor to go to the parent directory. A workaround is using the system back button, which will cause Markor to correctly go to the parent directory (if the last action you took was going to a child directory).
2. The "Download" symlink in "storage" and the "0" symlink in "storage/emulated" are crossed out and inaccessible. I didn't expect Markor to be able to access /storage/emulated/0/Download or /storage/emulated/0/ in the secondary profile. Instead, I expected Markor to have symlinks for /storage/emulated/10/Download and /storage/emulated/10.
3. Strange things happen when data is copied to "AppData (storage-emulated-#-Android-data-net.gsantner.markor)" (which I'll call "the public directory" from now on). Specifically:

* In Markor, all copied files and directories will appear in the public directory and can be accessed.
* In the system files app, the public directory usually appears to have nothing. Occasionally, some files that were copied will appear, but will be broken. If a directory was copied, that directory and its contents will not appear.
* Through MTP, the public directory will contain a "cache" directory and a "files" directory, as expected. Inside the "files" directory, some files that were copied will appear, but will be broken. If a directory was copied, the directory will appear, but it will appear to be empty.

When I say these files are broken and aren't accessible, I mean that trying to open them with another app doesn't work. My PC displays errors when attempting to read/copy/move/delete the files through MTP. This doesn't just happen with note files, this also happens with image files. 

Markor is storing the data *somewhere*. According to Android, storage gets taken up when Markor copies these broken files. I expected files and directories copied to the public directory to show up and be accessible by the system files app and through MTP.

***

Where this issue doesn't manifest:

* When using Markor in the system user. I'm not sure if Markor works correctly in other profiles (Private Space, Work, Restricted, Clone) or the Guest user.
* In Marder, copy operations to symlinks work correctly. Files copied to the public directory (/storage/emulated/#/Android/data/net.gsantner.markor_test) can be read by the system files app and through MTP. The other problems are still present in Marder, though.
* I don't remember this being an issue on the v2.12.# releases from July 2024. Though, I didn't test those when making this issue.

***

I want this bug resolved because most of my notes are stored in Markor's data partition in a secondary user. Resolving this issue would make it possible to copy files from the data partition to somewhere another computer can access the files via MTP, and it would restore intended behavior when browsing though the filesystem. Right now, in secondary users, it isn't possible to accees or make a backup copy of files in the data partition or the public directory, nor is it possible to enter /storage/emulated/#/ in Markor if you didn't already have save locations for the Notebook/To-Do/QuickNote set there.

### Steps to reproduce

1. In a secondary user, open Markor and enter a directory. Tap "..". You will be at "storage" and not the parent directory. To workaround this issue, open Markor and enter a directory. Then, you can use the system back button to go to the parent directory. However, if your last action was going to a symlink, then the system back button will take you there.
2. In a secondary user, open Markor and go to storage by tapping "..". Try tapping on "~~Download~~", or tap on "emulated" then try tapping on "~~0~~". 
3. In a secondary user, open Markor. Try copying a file and a directory with a file inside of it to the public directory. Then, try accessing the copied data in the system files app and through MTP. 

### Information

Android version: 15 (AP3A.241105.007)
Device: Google Pixel 7 Pro
Custom ROM: GrapheneOS
App Version: 2.13.1

### Source

GitHub Releases

### Format / File type

Not specific

### Additional info / Log

```shell
-
```

### PR DESCRIPTION
**Pull request by harshad1**
Supporting multiple users and adding a number of fixes

This addresses #2479

### CODE REVIEW

File: app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java [hunk 1]
```diff
@@ -230,10 +239,23 @@ public void onBindViewHolder(@NonNull FilesystemViewerViewHolder holder, int pos
...
+        if (isSelected) {
+            holder.image.setImageResource(isSelected ? _dopt.selectedItemImage : isFile ? _dopt.fileImage : _dopt.folderImage);
+        } else if (VIRTUAL_STORAGE_FAVOURITE.equals(displayFile)) {
+            holder.image.setImageResource(R.drawable.ic_star_black_24dp);
+        } else if (VIRTUAL_STORAGE_POPULAR.equals(displayFile)) {
+            holder.image.setImageResource(R.drawable.ic_favorite_black_24dp);
+        } else if (VIRTUAL_STORAGE_RECENTS.equals(displayFile)) {
+            holder.image.setImageResource(R.drawable.ic_history_black_24dp);
+        } else {
+            holder.image.setImageResource(isFile ? _dopt.fileImage : _dopt.folderImage);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java:251]: Use the specific images for these folders

File: app/src/main/java/net/gsantner/markor/frontend/filebrowser/MarkorFileBrowserFactory.java [hunk 2]
```diff
@@ -81,9 +81,8 @@ public static GsFileBrowserOptions.Options prepareFsViewerOpts(
             opts.favouriteFiles = appSettings.getFavouriteFiles();
             opts.recentFiles = appSettings.getRecentFiles();
             opts.popularFiles = appSettings.getPopularFiles();
-            opts.storageMaps.clear();
-            opts.storageMaps.put(new File("/storage", cu.rstr(context, R.string.notebook)), appSettings.getNotebookDirectory());
-            opts.storageMaps.put(new File("/storage/Download"), new File("/storage/emulated/0/Download"));
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/filebrowser/MarkorFileBrowserFactory.java:86]: Download folder resolved using API now

File: app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserFragment.java [hunk 3]
```diff
@@ -235,7 +235,7 @@ private void updateMenuItems() {
         // Check if is a favourite
         boolean selTextFilesOnly = true;
         boolean selDirectoriesOnly = true;
-        boolean selWritable = (!curFilepath.equals("/storage") && !curFilepath.equals("/storage/emulated"));
+        boolean selWritable = true;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserFragment.java:238]: We don't hardcode these like this now

File: app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java [hunk 4]
```diff
@@ -679,27 +707,28 @@ private void loadFolder(final File folder, final File show) {
         final File toShow = show == null ? _fileToShowAfterNextLoad : show;
         _fileToShowAfterNextLoad = null;
 
+        _currentFolder = folder;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java:710]: We set this early and in the wrapping sync function. Prevents bugs from preventing this from being set.

File: app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java [hunk 5]
```diff
@@ -653,24 +653,16 @@ public void onFsViewerConfig(GsFileBrowserOptions.Options dopt) {
                 return true;
             }
             case R.id.action_toggle_case:
-                if (_hlEditor != null) {
-                    _hlEditor.toggleCase();
-                }
+                TextViewUtils.toggleSelectionCase(_hlEditor.getText());
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/activity/DocumentEditAndViewFragment.java:656]: Cleanups to how these are structured and organized

File: app/src/main/java/net/gsantner/markor/frontend/filebrowser/MarkorFileBrowserFactory.java [hunk 6]
```diff
@@ -68,24 +70,24 @@ public static GsFileBrowserOptions.Options prepareFsViewerOpts(
...
-            opts.favouriteFiles = appSettings.getFavouriteFiles();
-            opts.recentFiles = appSettings.getRecentFiles();
-            opts.popularFiles = appSettings.getPopularFiles();
-            opts.storageMaps.clear();
-            opts.storageMaps.put(new File("/storage", cu.rstr(context, R.string.notebook)), appSettings.getNotebookDirectory());
-            opts.storageMaps.put(new File("/storage/Download"), new File("/storage/emulated/0/Download"));
-        };
-        opts.refresh.callback();
+        // These should be refreshed in onFsViewerConfig
+        opts.sortFolderFirst = appSettings.isFileBrowserSortFolderFirst();
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/filebrowser/MarkorFileBrowserFactory.java:77]: These don't really need to be refreshed dynamically

File: app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java [hunk 7]
```diff
@@ -186,69 +192,78 @@ public boolean isFileWriteable(File file, boolean isGoUp) {
...
-        } else if (_virtualMapping.containsKey(displayFile)) {
-            file = _virtualMapping.get(displayFile);
-        } else {
-            file = displayFile;
         }
 
-        final String filename = displayFile.getName();
-        final String currentFolderName = _currentFolder != null ? _currentFolder.getName() : "";
-        final File currentFolderParent = _currentFolder != null ? _currentFolder.getParentFile() : null;
+        final File file = GsCollectionUtils.getOrDefault(_virtualMapping, displayFile, displayFile);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java:201]: All of this is identical but cleaned up

File: app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java [hunk 8]
```diff
@@ -648,58 +642,48 @@ public boolean scrollToAndFlash(final File file) {
...
-            executorService.execute(() -> _loadFolder(toLoad, toShow));
+            executorService.execute(() -> _loadFolder(folderChanged, toShow));
         } catch (RejectedExecutionException ignored) { // during exit
         }
     }
 
-    // This function is not called on the main thread, so post to the UI thread
-    private synchronized void _loadFolder(final @NonNull File folder, final @Nullable File toShow) {
+    // This function is not called on the main thread
+    private synchronized void _loadFolder(final boolean folderChanged, final @Nullable File toShow) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/frontend/filebrowser/GsFileBrowserListAdapter.java:674]: We set _currentFolder early and just use that. Prevents some bugs

**Comment by harshad1** [comment]
I think this is good to go. But I will wait for feedback and more testing.

**Review by gsantner — APPROVED** [review]

**Comment by gsantner** [comment]
Thanks, merged

**Pull request merged**